sound/soc/sof/amd/acp-stream.c

```diff
@@ -0,0 +1,210 @@
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp4_pte);
+		stream->reg_offset = PTE_GRP4_OFFSET;
+		break;
+	case 5:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_5;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_5;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp5_pte);
+		stream->reg_offset = PTE_GRP5_OFFSET;
+		break;
+	case 6:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_6;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_6;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp6_pte);
+		stream->reg_offset = PTE_GRP6_OFFSET;
+		break;
+	case 7:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_7;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_7;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp7_pte);
+		stream->reg_offset = PTE_GRP7_OFFSET;
+		break;
+	case 8:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_8;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_8;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp8_pte);
+		stream->reg_offset = PTE_GRP8_OFFSET;
+		break;
+	default:
+		pr_err("Invalid stream tag %d\n", stream_tag);
+		break;
+	}
+
+	/* write phy_addr in scratch memory */
+
+	phy_addr_offset = offsetof(struct acp_scratch_register_config,
+				   reg_offset);
+	index = stream_tag - 1;
+	phy_addr_offset = phy_addr_offset + (index*4);
+
+	snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 +
+			  phy_addr_offset, stream->reg_offset);
+
+	/* Group Enable */
+	reg_val = ACP_SRAM_PTE_OFFSET + offset;
+	snd_sof_dsp_write(sdev, ACP_DSP_BAR, pte_reg, reg_val | BIT(31));
+	snd_sof_dsp_write(sdev, ACP_DSP_BAR, pte_size, PAGE_SIZE_4K_ENABLE);
+
+	for (page_idx = 0; page_idx < num_pages; page_idx++) {
+
+		addr = snd_sgbuf_get_addr(dmab, page_idx * PAGE_SIZE);
+
+		/* Load the low address of page int ACP SRAM through SRBM */
+		low = lower_32_bits(addr);
+		high = upper_32_bits(addr);
+
+		snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + offset, low);
+
+		high |= BIT(31);
+		snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_SCRATCH_REG_0 + offset + 4, high);
+		/* Move to next physically contiguous page */
+		offset += 8;
+	}
+
+	return 0;
+}
+
+struct acp_dsp_stream *
+acp_dsp_stream_get(struct snd_sof_dev *sdev)
+{
+	struct acp_dev_data *adata = sdev->pdata->hw_pdata;
+	struct acp_dsp_stream *stream;
+	int node = 0;
+
+	/* get an unused stream */
+	list_for_each_entry(stream, &adata->stream_list, list) {
+		node++;
+		if (stream->active == 0)
+			break;
+	}
+
+	stream->active = 1;
+	return stream;
+}
+EXPORT_SYMBOL_NS(acp_dsp_stream_get, SND_SOC_SOF_AMD_COMMON);
+
+int acp_dsp_stream_put(struct snd_sof_dev *sdev,
+		       struct acp_dsp_stream *acp_stream)
+{
+	struct acp_dev_data *adata = sdev->pdata->hw_pdata;
+	struct acp_dsp_stream *stream;
+	int ret = 0;
+
+	/* Free an active stream */
+	list_for_each_entry(stream, &adata->stream_list, list) {
+		if (stream == acp_stream)
+			break;
+	}
+
+	if (stream->active == 1) {
```

Comment:
is this check really necessary? can you not set it to 0 unconditionally? its a put function right?